volume/driver/overlay_linux.go

```diff
@@ -130,7 +127,12 @@ func (driver *OverlayDriver) Recover(fs volume.Filesystem) error {
 	}
 
 	for _, cow := range cows {
-		err = driver.overlayMount(cow.child, cow.parent)
+		rootParent, err := driver.findRootParent(cow.child, cow.parent)
+		if err != nil {
+			return err
+		}
+
+		err = driver.overlayMount(cow.child, rootParent)
```

Comment:
If I understand correctly, the `driver.Recover()` is called at startup to potentially restore the state of a previously-running-baggageclaim. For most other drivers, this is a noop (naive is understandable, but why is btrfs also noop?). But for overlayfs, the underlying volumes are accessible and contain enough symlink/metadata/reference/magic to reconstruct the entire hierarchy.

So my question is, under what circumstance would the parent of a volume not be the root parent, and why is it important we mount it to the root instead of immediate parent?

Reply:
> under what circumstance would the parent of a volume not be the root parent

When you make a COW (copy-on-write) volume whose parent is a COW volume. The real root volume is the parent of the second COW volume.

This commit explains it in more detail: https://github.com/concourse/baggageclaim/commit/0a8067418aede613d4b7431b7d5a7b34c697f5a1